index.ts

```diff
@@ -0,0 +1,106 @@
+import { FastifyPluginAsync } from "fastify";
+import { sql } from "slonik";
+import { Item, Member, PreHookHandlerType } from "graasp";
+
+const DEFAULT_MAX_STORAGE = 1024 * 1024 * 100; // 100MB;
+
+interface GraaspFileUploadLimiterOptions {
+  /** Item type to target (ex: 'file', 's3File') */
+  type: string;
+  /** Chain of property names to the file size (ex: 'extra.s3File.size')  */
+  sizePath: string;
+  /** Maximum storage size for a user in bytes (default to 100MB)  */
+  maxMemberStorage?: number;
+}
+
+type MemberExtra = {
+  storage: {
+    total: number;
+  };
+};
+
+class StorageExceeded extends Error {
+  message = "The allowed storage is full";
+  name = "StorageExceededError";
+  origin = "plugin";
+  code = 507;
+}
+
+const plugin: FastifyPluginAsync<GraaspFileUploadLimiterOptions> = async (fastify, options) => {
+  const {
+    items: { taskManager },
+    taskRunner: runner,
+    db,
+  } = fastify;
+  const { type: itemType, sizePath, maxMemberStorage = DEFAULT_MAX_STORAGE } = options;
+
+  if (!itemType || !sizePath) {
+    throw new Error("graasp-file-upload-limiter: missing plugin options");
+  }
+
+  const getFileSize = (item): number => {
+    const properties = sizePath.split(".");
+    const sizeField = properties.reduce((prev, curr) => prev?.[curr], item);
+
+    if (Number.isInteger(sizeField)) {
+      return sizeField;
+    }
+
+    throw new Error("sizePath does not lead to a size number");
+  };
+
+  const getMemberStorage = (id: string) => {
+    // todo: fetch total storage from Redis
+
+    // -- the following solution is failing because it cannot extract property from jsonb
+    // const column = sql.join(["extra", "file", "size"], sql`->`);
+    // return db.pool
+    //   .oneFirst<string>(
+    //     sql`
+    //        SELECT
+    //        SUM((${column})) as size
+    //        FROM item
+    //        WHERE item.type ='file' AND item.creator = ${id}`
+    //   )
+    //   .then((size) => parseInt(size, 10) || 0);
```

Comment:
I couldn't use sql `SUM` since I couldn't select jsonb properties using slonik (line 56 results in an error)... if you have any idea it would be great. Instead I did the sum using javascript.

Reply:
I managed to do it:
`const propertiesPath = sql.join(['file', 'size'], sql`->`);`

You should take out the `extra` from the join, because it's the column name, and just keep the json property names. Then in the query you do:
```
           SELECT
           SUM((extra->${propertiesPath})::int)
           FROM item
           (etc...)
```